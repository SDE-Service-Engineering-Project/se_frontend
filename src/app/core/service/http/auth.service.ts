import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {AuthResponse} from "../../model/AuthResponse";

const AUTH_API = 'http://localhost:8080/auth/';

@Injectable({
  providedIn: 'root'
})
export class AuthService {

  constructor(private http: HttpClient) {
  }

  public login(userName: string, password: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(
      AUTH_API + 'login',
      {userName, password}
    );
  }
}
