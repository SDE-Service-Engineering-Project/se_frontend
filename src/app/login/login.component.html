<div class="container-fluid center-screen">
  <div id="main-wrapper" class="container">
    <div class="row justify-content-center">
      <div class="col-xl-10">
        <div class="card border-0">
          <div class="card-body p-0">
            <div class="row no-gutters">
              <div class="col-lg-6">
                <div class="p-5">
                  <div class="mb-5">
                    <h3 class="h4 font-weight-bold text-theme" style="color: #5369f8">Login</h3>
                  </div>
                  <h6 class="h5 mb-0">Welcome back!</h6>
                  <p class="text-muted mt-2 mb-5">Enter your user name and password to access the best cars in the
                    world.</p>

                  <form [formGroup]="loginForm" id="loginForm" (ngSubmit)="submitLogin()" >
                    <div class="form-group" style="text-align: left">
                      <label for="inputUserName">Username</label>
                      <input type="text" class="form-control" id="inputUserName" formControlName="username"
                             [ngClass]="{ 'is-invalid': loginForm.get('username')?.touched && loginForm.get('username')!!.errors, 'is-valid': loginForm.get('username')!!.valid }"
                      >
                      <div *ngIf="loginForm.get('username')!!.touched && loginForm.get('username')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="loginForm.get('username')!!.errors!!.hasOwnProperty('required')" id="err-username">Username is required!
                        </div>
                      </div>
                    </div>

                    <div class="form-group mb-5 mt-2" style="text-align: left">
                      <label for="inputPassword">Password</label>
                      <input type="password" class="form-control" id="inputPassword" formControlName="password"
                             [ngClass]="{ 'is-invalid': loginForm.get('password')!!.touched && loginForm.get('password')!!.errors, 'is-valid': loginForm.get('password')!!.valid }"
                      >
                      <div *ngIf="loginForm.get('password')!!.touched && loginForm.get('password')!!.errors"
                           class="invalid-feedback">
                        <div *ngIf="loginForm.get('password')!!.errors!!.hasOwnProperty('required')" id="err-password">Password is
                          required!
                        </div>
                      </div>
                    </div>
                    <div class="form-group mb-5">
                      <button type="submit" class="btn btn-theme form-control" id="login-btn" [disabled]="loginForm!!.invalid">Login</button>
                    </div>
                  </form>
                </div>
              </div>
              <div class="col-lg-6 d-none d-lg-inline-block">
                <div class="logo rounded-right">
                  <div class="overlay rounded-right"></div>
                  <div class="logo-inner">
                    <h4 class="text-white mb-4">Welcome to Car Rental</h4>
                    <p class="lead text-white">"Where cars are still cars."</p>
                    <p>- Godfather (CEO)</p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <!--TODO: add register component-->
        <p class="text-muted text-center mt-3 mb-0">Don't have an account? Register <a
          href="https://www.youtube.com/watch?v=dQw4w9WgXcQ" target="_blank" rel="noopener" class="text-primary ml-1">here</a>.
        </p>
      </div>
    </div>
  </div>
</div>
