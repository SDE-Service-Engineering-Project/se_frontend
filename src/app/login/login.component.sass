.logo
  padding: 0
  background-image: url(https://media.tenor.com/d0BeKMg4ohYAAAAC/car.gif)
  background-repeat: no-repeat
  background-size: cover
  height: 100%
  position: relative
  .overlay
    -webkit-box-flex: 1
    -ms-flex: 1
    flex: 1
    position: absolute
    top: 0
    bottom: 0
    left: 0
    right: 0
    background-color: rgba(0, 0, 0, 0.4)

  .logo-inner
    text-align: center
    color: #fff
    position: absolute
    margin: 0 auto
    padding: 0 1.75rem
    bottom: 3rem
    left: 0
    right: 0

.btn-theme
  background-color: #5369f8
  border-color: #5369f8
  color: #fff

.center-screen
  display: flex
  justify-content: center
  align-items: center
  text-align: center
  min-height: 100vh
