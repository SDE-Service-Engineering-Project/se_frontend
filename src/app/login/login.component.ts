import {Component, OnInit} from '@angular/core';
import {FormControl, FormGroup, Validators} from "@angular/forms";

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.sass']
})
export class LoginComponent implements OnInit {
  loginForm = new FormGroup({
    username: new FormControl("", [Validators.required]),
    password: new FormControl("", [Validators.required])
  });


  ngOnInit(): void {

  }


  submitLogin() {
    if (this.loginForm.invalid) return;

    console.log("make login call user:" + this.loginForm.get('username')!!.value + " password: " + this.loginForm.get('password')!!.value)
  }
}
