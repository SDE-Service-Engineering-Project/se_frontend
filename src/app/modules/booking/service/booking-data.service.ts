import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Booking } from '../../../models/booking';

export const BOOKING_API = 'http://localhost:8080/api/v1/bookings';

@Injectable({
  providedIn: 'root',
})
export class BookingDataService {
  constructor(private http: HttpClient) {}

  fetchBookings(): Observable<Booking[]> {
    return this.http.get<Booking[]>(BOOKING_API);
  }
}
