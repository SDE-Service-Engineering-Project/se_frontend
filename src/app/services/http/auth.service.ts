import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {AuthResponse} from '../../models/AuthResponse';

export const AUTH_API = 'http://localhost:8080/api/v1/auth/';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  constructor(private http: HttpClient) {
  }

  public login(userName: string, password: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(AUTH_API + 'login', {
      userName,
      password,
    });
  }

  public register(userName: string, firstName: string, lastName: string, password: string): Observable<any> {
    return this.http.post(AUTH_API + 'signup', {
      userName,
      firstName,
      lastName,
      password,
    });
  }

  public refreshToken(refreshToken: string, userName: string): Observable<AuthResponse> {
    return this.http.post<AuthResponse>(AUTH_API + 'refresh/token', {
      refreshToken,
      userName,
    });
  }
}
