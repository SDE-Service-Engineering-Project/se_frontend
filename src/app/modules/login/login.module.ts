import { NgModule } from '@angular/core';
import { LoginComponent } from './login.component';
import { LoginRoutingModule } from './config/login.routing';

@NgModule({
  imports: [LoginRoutingModule],
  declarations: [LoginComponent],
})
export class LoginModule {}
