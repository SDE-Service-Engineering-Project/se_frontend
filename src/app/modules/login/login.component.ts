import {Component, OnInit} from '@angular/core';
import {FormControl, FormGroup, Validators} from '@angular/forms';
import {AuthService} from '../../services/http/auth.service';
import {StorageService} from '../../services/storage.service';
import {ToastService} from '../../services/toast/toast.service';
import {Router} from "@angular/router";

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.sass'],
})
export class LoginComponent implements OnInit {
  loginForm = new FormGroup({
    username: new FormControl('', [Validators.required]),
    password: new FormControl('', [Validators.required]),
  });

  constructor(
    private authService: AuthService,
    private storageService: StorageService,
    private toastService: ToastService,
    private router: Router
  ) {
  }

  ngOnInit(): void {
  }

  submitLogin() {
    if (this.loginForm.invalid) return;

    this.toastService.removeAll();

    this.authService
      .login(this.loginForm.get('username')?.value!!, this.loginForm.get('password')?.value!!)
      .subscribe({
        next: (data) => {
          this.storageService.saveToken(data);
          this.router.navigate(['/home'])
            .then(() => this.toastService.showDefaultSuccessToast("Login successfully"));
        },
        error: (err) => {
          this.loginForm.get('password')?.setValue('');
          this.toastService.showDefaultErrorToast(err.error.message);
        },
      });
  }


}
