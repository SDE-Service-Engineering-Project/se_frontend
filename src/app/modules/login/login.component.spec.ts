import { ComponentFixture, TestBed } from '@angular/core/testing';

import { LoginComponent } from './login.component';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { AuthService } from '../../services/http/auth.service';
import { StorageService } from '../../services/storage.service';

describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;
  let spectator: Spectator<LoginComponent>;

  const createComponent = createComponentFactory({
    detectChanges: true,
    component: LoginComponent,
    imports: [ReactiveFormsModule, HttpClientModule],
    providers: [AuthService, StorageService],
  });

  beforeEach(() => {
    spectator = createComponent();
    fixture = spectator.fixture;
    component = spectator.component;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should display login button', function () {
    const loginButton =
      fixture.debugElement.nativeElement.querySelector('#login-btn');
    expect(loginButton).toBeDefined();
  });

  it('should return valid login form is username is set', function () {
    component.loginForm.get('username')?.setValue('username');

    expect(component.loginForm.get('username')?.valid).toBeTruthy();
    expect(
      fixture.debugElement.nativeElement.querySelector('#err-username')
    ).toBeNull();
  });

  it('should return invalid login form if username is empty', function () {
    component.loginForm.get('username')?.setValue('');
    expect(component.loginForm.get('username')?.invalid).toBeTruthy();
    expect(
      fixture.debugElement.nativeElement.querySelector('#err-username')
    ).toBeDefined();
  });

  it('should return valid login form if password is set', function () {
    component.loginForm.get('password')?.setValue('myAwesomePassword');
    expect(component.loginForm.get('password')?.valid).toBeTruthy();
    expect(
      fixture.debugElement.nativeElement.querySelector('#err-password')
    ).toBeNull();
  });

  it('should return invalid login form if password is not set', function () {
    component.loginForm.get('password')?.setValue('');
    expect(component.loginForm.get('password')?.invalid).toBeTruthy();
    expect(
      fixture.debugElement.nativeElement.querySelector('#err-password')
    ).toBeDefined();
  });

  it('should enable login button when username and password is set', function () {
    expect(
      fixture.debugElement.nativeElement.querySelector('#login-btn').disabled
    ).toBeTruthy();
    component.loginForm.get('username')?.setValue('username');
    component.loginForm.get('password')?.setValue('password');

    fixture.detectChanges();

    expect(
      fixture.debugElement.nativeElement.querySelector('#login-btn').disabled
    ).toBeFalsy();
  });

  it('should not enable login button when only username is set', function () {
    expect(
      fixture.debugElement.nativeElement.querySelector('#login-btn').disabled
    ).toBeTruthy();
    component.loginForm.get('username')?.setValue('username');

    fixture.detectChanges();

    expect(
      fixture.debugElement.nativeElement.querySelector('#login-btn').disabled
    ).toBeTruthy();
  });
});
