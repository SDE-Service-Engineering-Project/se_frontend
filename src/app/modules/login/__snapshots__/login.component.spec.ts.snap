// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should render the component 1`] = `
<div
  id="root0"
>
  <div
    class="container-fluid center-screen"
  >
    <div
      class="container"
      id="main-wrapper"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-xl-10"
        >
          <div
            class="card border-0"
          >
            <div
              class="card-body p-0"
            >
              <div
                class="row no-gutters"
              >
                <div
                  class="col-lg-6"
                >
                  <div
                    class="p-5"
                  >
                    <div
                      class="mb-5"
                    >
                      <h3
                        class="h4 font-weight-bold text-theme"
                        style="color: #5369f8;"
                      >
                        Login
                      </h3>
                    </div>
                    <h6
                      class="h5 mb-0"
                    >
                      Welcome back!
                    </h6>
                    <p
                      class="text-muted mt-2 mb-5"
                    >
                      Enter your user name and password to access the best cars in the world.
                    </p>
                    <form
                      class="ng-untouched ng-pristine ng-invalid"
                      id="loginForm"
                      novalidate=""
                    >
                      <div
                        class="form-group"
                        style="text-align: left;"
                      >
                        <label
                          for="inputUserName"
                        >
                          Username
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="username"
                          id="inputUserName"
                          type="text"
                        />
                        
                      </div>
                      <div
                        class="form-group mb-5 mt-2"
                        style="text-align: left;"
                      >
                        <label
                          for="inputPassword"
                        >
                          Password
                        </label>
                        <input
                          class="form-control ng-untouched ng-pristine ng-invalid"
                          formcontrolname="password"
                          id="inputPassword"
                          type="password"
                        />
                        
                      </div>
                      <div
                        class="form-group mb-5"
                      >
                        <button
                          class="btn btn-theme form-control"
                          data-testid="login-btn"
                          disabled=""
                          id="login-btn"
                          type="submit"
                        >
                          Login 
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
                <div
                  class="col-lg-6 d-none d-lg-inline-block"
                >
                  <div
                    class="logo rounded-right"
                  >
                    <div
                      class="overlay rounded-right"
                    />
                    <div
                      class="logo-inner"
                    >
                      <h4
                        class="text-white mb-4"
                      >
                        Welcome to Car Rental
                      </h4>
                      <p
                        class="lead text-white"
                      >
                        "I wanted to give the folks a little sizzle. Ka-chow!"
                      </p>
                      <p>
                        - Lightning McQueen
                      </p>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <p
            class="text-muted text-center mt-3 mb-0"
          >
            Don't have an account? Register 
            <a
              class="text-primary ml-1"
              href="/register"
              rel="noopener"
              routerlink="/register"
            >
              here
            </a>
            . 
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
